core/server/master/src/main/java/alluxio/master/file/mdsync/LoadRequestExecutor.java

```diff
@@ -0,0 +1,269 @@
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Map;
+import java.util.Optional;
+import java.util.PriorityQueue;
+import java.util.Set;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+import javax.annotation.Nullable;
+
+class LoadRequestExecutor implements Closeable {
+  private static final Logger LOG = LoggerFactory.getLogger(LoadRequestExecutor.class);
+  private static final Logger SAMPLING_LOG = new SamplingLogger(LOG, 5L * Constants.SECOND_MS);
+
+  /** Limit the number of running (or completed but not yet processed) load requests. **/
+  private final AtomicInteger mRemainingTickets;
+  private final int mMaxRunning;
+
+  private final Map<Long, PathLoaderTask> mPathLoaderTasks = new ConcurrentHashMap<>();
+  // Loader tasks with pending loads
+  private final Set<Long> mPathLoaderTasksWithPendingLoads = new ConcurrentHashSet<>();
+  // Same as above, except ordered by priority
+  private final ConcurrentLinkedDeque<Long> mPathLoaderTaskQueue = new ConcurrentLinkedDeque<>();
+  // Load requests in order of to be processed
+  private final BlockingQueue<LoadRequest> mLoadRequests = new LinkedBlockingQueue<>();
+  // Rate limited loads that are not yet ready to be run
+  private final PriorityQueue<RateLimitedRequest> mRateLimited = new PriorityQueue<>();
+
+  private final LoadResultExecutor mResultExecutor;
+
+  private final Thread mExecutor;
+
+  LoadRequestExecutor(int maxRunning, LoadResultExecutor resultExecutor) {
+    mMaxRunning = maxRunning;
+    mRemainingTickets = new AtomicInteger(maxRunning);
+    mResultExecutor = resultExecutor;
+    mExecutor = new Thread(() -> {
+      while (!Thread.interrupted()) {
+        try {
+          runNextLoadTask();
+        } catch (InterruptedException e) {
+          return;
+        }
+      }
+      LOG.info("Load request runner thread exiting");
+    }, "LoadRequestRunner");
+    mExecutor.start();
+    registerMetrics();
+  }
+
+  synchronized void addPathLoaderTask(PathLoaderTask task) {
+    long id = task.getTaskInfo().getId();
+    task.runOnPendingLoad(() -> hasNewLoadTask(id));
+    mPathLoaderTasks.put(id, task);
+    mPathLoaderTaskQueue.add(id);
+    mPathLoaderTasksWithPendingLoads.add(id);
+    notifyAll();
+  }
+
+  synchronized void hasNewLoadTask(long taskId) {
+    if (!mPathLoaderTasksWithPendingLoads.contains(taskId)) {
+      mPathLoaderTaskQueue.add(taskId);
+      mPathLoaderTasksWithPendingLoads.add(taskId);
+      notifyAll();
+    }
+  }
+
+  private void onLoadError(LoadRequest request, Throwable t) {
+    // TODO(elega) this might result in load request that is retried successfully being
+    // added incorrectly
+    if (t instanceof DefaultSyncProcess.MountPointNotFoundRuntimeException) {
```

Comment:
@elega check please?

Reply:
yeah i revised the comment a bit 
this is actually not a TODO